tools/README.md

```diff
@@ -0,0 +1,193 @@
+- `-H`: pass this flag is you want to only measure the timing iteration of a benchmark. By default, the tracing tools will measure the entire execution.
+- `-p`: print the entire tracing script before execution. This is mainly for debugging use.
+- `-f`: change the bpftrace output format. By default, it uses human-readable plain text output (`text`). You can set this to `json` for easy parsing.
+
+Please run the tracing tools **before** running the workload. If you use `-H`, the tracing tools will automatically end with `harness_end` is called. Otherwise, you will need to terminate the tools manually with `Ctrl-C`. These tools also have a timeout of 1200 seconds so not to stall unattended benchmark execution. 
+
+## Tracing tools
+### Measuring the time spend in allocation slow path (`alloc_slow`)
+This tool measures the distribution of the allocation slow path time.
+The time unit is ns / 400, so that we use the histogram bins with higher fidelity better.
+
+Sample output:
+```
+@alloc_slow_hist: 
+[4, 8)               304 |@                                                   |
+[8, 16)            12603 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@|
+[16, 32)            8040 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@                   |
+[32, 64)             941 |@@@                                                 |
+[64, 128)            171 |                                                    |
+[128, 256)            13 |                                                    |
+[256, 512)             2 |                                                    |
+[512, 1K)              0 |                                                    |
+[1K, 2K)               0 |                                                    |
+[2K, 4K)               0 |                                                    |
+[4K, 8K)               0 |                                                    |
+[8K, 16K)              0 |                                                    |
+[16K, 32K)            14 |                                                    |
+[32K, 64K)            37 |                                                    |
+[64K, 128K)           19 |                                                    |
+[128K, 256K)           1 |                                                    |
+```
+
+In the above output, we can see that most allocation slow paths finish between 3.2us and 6.4us. However, there is a long tail, presumably due to GC pauses.
+
+### Measuring the time spend in different GC stages (`gc_stages`)
+This tool measures the time spent in different stages of GC: before `Closure`, during `Closure`, and after `Closure`.
+The time unit is ns.
```

Comment:
A general style problem: I find that you never uses newline characters within one paragraph.  Markdown allows this.

But here the text spans over three lines.  I think you intend to introduce one title and two paragraphs.  However, Markdown will interpret consecutive non-blank lines as one paragraph (except for titles (`### blah blah blah`)).  If you intend to put "The time unit is ns." in a single paragraph, you need to add an empty line before it.

I also recommend formatting the text so that the line width is no more than 100 characters long (or 80 or 72 if you prefer).  The `fmt` command line utility may come in handy.  If you are a Vim user, you can `set tw=100`, select one or more paragraphs and press `gq` to format them.  In this way, this text file will still be easy to read using text editors.

Reply:
Done